grease/src/Grease/Shape.hs

```diff
@@ -314,3 +338,206 @@ deriving instance (ShowExt ext tag, ShowF tag) => Show (ArgShapes ext tag tys)
+  -- need to compute padding between els, infront of els and at end
+  shapeSeq (MDwarf.StructType sdecl) =
+    let structSize = MDwarf.structByteSize sdecl
+     in do
+          (currLoc, seqs) <-
+            foldM
+              ( \(currloc, seqs) mem ->
+                  do
+                    (nextLoc, additionalShapes) <- buildMember currloc mem
+                    Just (nextLoc, seqs Seq.>< additionalShapes)
+              )
+              (0, Seq.empty)
+              (MDwarf.structMembers sdecl)
+          endPad <- Just (if currLoc >= structSize then Seq.empty else Seq.singleton (padding $ structSize - currLoc))
+          Just $ (seqs Seq.>< endPad)
+  shapeSeq _ = Nothing
+
+intPtrShape :: Symbolic.SymArchConstraints arch => MC.ArchReg arch tp -> Maybe (C.Some (PtrShape ext w NoTag))
+intPtrShape reg =
+  case MT.typeRepr reg of
+    MT.BVTypeRepr w -> Just $ C.Some $ ShapePtrBV NoTag w
+    _ -> Nothing
+
+pointerShapeOfDwarf :: (HasPtrWidth w, Symbolic.SymArchConstraints arch) => ArchContext arch -> MC.ArchReg arch tp -> Subprogram -> MDwarf.TypeApp -> Maybe (C.Some (PtrShape ext w NoTag))
+pointerShapeOfDwarf _ r _ (MDwarf.SignedIntType _) = intPtrShape r
+pointerShapeOfDwarf _ r _ (MDwarf.UnsignedIntType _) = intPtrShape r
+pointerShapeOfDwarf _ _ sprog (MDwarf.PointerType _ tyRef) =
+  let mshape = constructPtrTarget sprog =<< extractType sprog =<< tyRef
+      pshape = ShapePtr NoTag (Offset (toBytes (0 :: Int))) <$> mshape
+   in (C.Some <$> pshape)
+pointerShapeOfDwarf _ _ _ _ = Nothing
+
+-- Is there really nothing available that looks like this?
+takeJust :: (a -> Maybe b) -> [a] -> [b]
+takeJust _ [] = []
+takeJust f (h : tl) =
+  case f h of
+    Nothing -> []
+    Just e -> e : takeJust f tl
+
+shapeFromVar :: (ExtShape ext ~ PtrShape ext wptr, Mem.HasPtrWidth wptr, Symbolic.SymArchConstraints arch) => ArchContext arch -> MC.ArchReg arch tp -> Subprogram -> MDwarf.Variable -> Maybe (C.Some (Shape ext NoTag))
+shapeFromVar arch buildingForReg sprog vr =
+  C.mapSome ShapeExt <$> (pointerShapeOfDwarf arch buildingForReg sprog =<< extractType sprog =<< MDwarf.varType vr)
+
+shapeFromDwarf :: (Symbolic.SymArchConstraints arch, ExtShape ext ~ PtrShape ext wptr, Mem.HasPtrWidth wptr) => ArchContext arch -> Subprogram -> ParsedShapes ext
+shapeFromDwarf aContext sub =
+  let
+    abiregs = aContext Lens.^. archABIParams
+    args = (zip abiregs $ snd <$> (toAscList $ subParamMap sub))
+    ascParams =
+      takeJust
+        ( \(reg, var) ->
+            do
+              C.Some r <- pure reg
+              shp <- shapeFromVar aContext r sub var
+              pure (Text.pack $ coerce $ mkRegName r, shp)
+        )
+        args
+    named = Map.fromList $ ascParams
+   in
+    ParsedShapes{_getParsedShapes = named}
+
+fromDwarfInfo :: (Symbolic.SymArchConstraints arch, ExtShape ext ~ PtrShape ext wptr, Mem.HasPtrWidth wptr) => ArchContext arch -> Word64 -> [Data.Macaw.Dwarf.CompileUnit] -> Maybe (ParsedShapes ext)
```

Comment:
Style suggestion: this line is quite long, and it would be helpful to split it up into multiple lines for the sake of readability. Something like this, perhaps:

```hs

fromDwarfInfo ::
  ( Symbolic.SymArchConstraints arch
  , ExtShape ext ~ PtrShape ext wptr
  , Mem.HasPtrWidth wptr
  ) =>
  ArchContext arch ->
  Word64 ->
  [Data.Macaw.Dwarf.CompileUnit] ->
  Maybe (ParsedShapes ext)
```

And similarly for the other newly introduced functions in this module.